### ISSUE TITLE
hdf5 fails to build with Xcode 15 / macOS Sonoma

### ISSUE THREAD

**Issue by fxcoudert** [HDFGroup/hdf5#3571]
**Describe the bug**

Compilation with Xcode 15 (on macOS Sonoma) leads to a build error:

```
  ld: unknown options: -commons 
  clang: error: linker command failed with exit code 1 (use -v to see invocation)
  make[1]: *** [H5detect] Error 1
```

Apple ships with Xcode 15 and macOS Sonoma a new linker, which does not accept the `-commons` option.

This commons from hdf5 configure at https://github.com/HDFGroup/hdf5/blob/95907ba548cd6d40eb625f7c79d9bc437df93c68/configure.ac#L1082

```
  ## tell libtool to do the right thing with COMMON symbols, this fixes
  ## corrupt values with COMMON and EQUIVALENCE when building shared
  ## Fortran libraries on OSX with gnu and Intel compilers (HDFFV-2772).
  case "`uname`" in
    Darwin*)
    H5_LDFLAGS="$H5_LDFLAGS -Wl,-commons,use_dylibs"
    ;;
  esac
```

**Expected behavior**
Successful install :)

**Platform (please complete the following information)**
 - HDF5 version: 1.14.2
 - OS and version: macOS 14.0 (Release Candidate)
 - Compiler and version: Xcode 15
 - Build system (e.g. CMake, Autotools) and version: autoconf 2.71, automake 1.66.5, libtool 2.4.7
 - Any configure options you specified: `./configure --disable-dependency-tracking --disable-silent-rules --enable-build-mode=production --enable-fortran --enable-cxx --prefix=/opt/homebrew/Cellar/hdf5/1.14.2 --with-szlib=/opt/homebrew/opt/libaec`
 - MPI library and version (parallel HDF5): MPI is not used

**Additional context**

**Comment by fxcoudert** [comment]
I've filed the bug with Apple as FB13194355

**Comment by Dave-Allured** [comment]
@fxcoudert, thank you for the report.  Macports users are now running into this, unknown `-commons`, after updating to Xcode 15 and/or Command Line Tools 15, since 15.0 release three days ago (2023 September 18).
https://trac.macports.org/ticket/68194

**Comment by Dave-Allured** [comment]
Note:  `-commons` is a fortran-related option.  A Macports user has reported that HDF5 1.14.2 without the fortran API option builds fine with Xcode 15.0.

**Comment by Dave-Allured** [comment]
@fxcoudert, can you test with `--enable-fortran` removed (or replaced by `--disable-fortran`)?

**Comment by Dave-Allured** [comment]
History:  `-commons` was added on 2017 May 23.  It has been present in all HDF5 releases since 1.8.19 and 1.10.2.  This commit:
https://github.com/HDFGroup/hdf5/commit/133168d8fcd5210c66d65ba3c1b349f72fd39f0f

**Comment by fxcoudert** [comment]
@Dave-Allured we build HDF5 with Fortran support in Homebrew. We've worked around the issue by forcing use of the old linker: https://github.com/Homebrew/homebrew-core/pull/144559

**Comment by Dave-Allured** [comment]
@fxcoudert, thanks.  I will see if that can be worked into Macports.

**Comment by brtnfld** [comment]
It also will fail with Fortran disabled but with parallel enabled. At least with mpich, since it includes the -commons in the mpicc wrapper.

**Comment by Dave-Allured** [comment]
`-commons` is an obscure option that is found only in Mac OS linkers.  As cited above:
 ```
## ... this fixes corrupt values with COMMON and EQUIVALENCE when
## building shared Fortran libraries on OSX with gnu and Intel compilers
```
I searched several versions of HDF5 source code.  **COMMON** and **EQUIVALENCE** were used only in fortran, and only in the 1.8 release series (and maybe earlier), lately in the single file **H5f90global.f90**.

I believe that `-commons` is now completely unnecessary for building series 1.10 forward, including fortran.  Therefore I recommend removing `-commons` in all library patches and future releases.  This is in preference to down-shifting to the old linker.  I have not yet tested anywhere to see if this really works.

Would someone at HDF5 please verify this finding?  Thank you.

User programs that use **COMMON** or **EQUIVALENCE** will probably still need to compile and link with `-commons` on Mac.

**Comment by brtnfld** [comment]
You are correct, it was overlooked when we removed the use of COMMON and EQUIVALENCE in the switch to Fortran 2003. Thanks.

**Comment by derobins** [comment]
Will also be fixed in 1.10.11 and 1.12.3
